### DISCUSSION TITLE
What do you guys think about KMP?

### DISCUSSION THREAD

**Discussion by sanjeevirajm** [skiptools/skip#6] [Q&A] [answered]
I saw the below page
https://skip.tools/compare/

Few disadvantages in KMP when compared to Skip - 
* It uses Garbage collection and not ARC
* Even though swift interop is in roadmap it is not available right now

And another major difference is platform bias. Skip towards iOS, KMP towards Android.
For a team with equal no of Swift developers and Kotlin developers, which tool would you recommend? 
Skip or KMP. 
If Skip, is there any advantages compared to KMP other than anything mentioned in the comparison page

**Comment by sanjeevirajm** [comment]
Writing a KMP SDK is easy.
The SDK can touch lots of platform specific API's via separate iOS source set and Android source set.
how will it be in Skip?

**Comment by marcprux** [comment, marked as answer]
The advantages you list are big ones: with Skip you don't have an embedded garbage collector slowing down and bloating the iOS side of your project, and only Skip gives you true Swift interoperability. The latter advantage will only become more acute as iOS continues to release Swift-only frameworks, as all the other frameworks can only work with Swift via the (limited) Objective-C bridge. 

For example, only Skip can work directly with SwiftUI – all the other frameworks need to bridge through Objective-C to UIKit if they want to render native components.

Another advantage of Skip is that our use of the platform-blessed UI framework (SwiftUI and Compose, respectively) means that it is impossible for a Skip app to "lag behind" platform updates in terms of UI improvements and other changes in behavior. Your Skip app will always feel as genuinely native as the apps that are shipped with the operating system itself.

To your question about library development: this is an area where Skip really shines. Our library development workflow enables you to test against both platforms simultaneously, using Robolectric for Android API testing. This works both when running from Xcode, and also headlessly, enabling you to maintain separate CI pipelines for each of your individual modules. Read more about it https://skip.tools/docs/modules/skip-unit/

What it really comes down to, though, is which development platform makes you happiest and most productive. If you prefer Kotlin, or are already heavily invested in the Kotlin ecosystem, then KMM/KMP may be the best choice for you. Similarly, if you are in love with the Dart programming language, then Flutter is the obvious choice. And there are lots of other choices if JavaScript is your preferred language for app development.

We love Swift – and the development productivity that comes from building, testing, and running your app all in Xcode – which is why we invented Skip.

**Reply by sanjeevirajm** [reply]
Thanks for the insights

**Comment by sanjeevirajm** [comment]
is it possible to share only few things in Skip across iOS and Android? 
Say there is an existing app available, wanted to write a new feature alone in Skip and share business logic alone.
is that possible? How the interop will work in Kotlin?

**Reply by marcprux** [reply]
Skip transpiles your SPM modules into Gradle modules. You can depend on these Gradle modules from a separate Kotlin project, either on the source level, or by simply linking to their output artifacts (e.g., like we publish for our core modules like https://source.skip.tools/skip-foundation/releases). 

You may need to manually import the Gradle project into your other Kotlin project(s), but once you've done so, you can call your transpiled Swift code just like any other Kotlin code. 

Skip's transpiled Kotlin code may contains a few idiosyncrasies, but in general should feel quite fluent to a Kotlin developer. You can explore it on our [Skip Playground](https://skip.tools/playground/).
